Fix legalization of stablehlo.dot_general to TTIR matmul

LPanosTT opened the pull request:
Issue: https://github.com/tenstorrent/tt-mlir/issues/1047

Broaden scope of legalization for `stablehlo.dot_general` to `ttir.matmul` to capture all cases where they perform the same computation.

mmanzoorTT commented:
Please add tests in `test/ttmlir/Silicon/StableHLO/` to ensure that `dot_general` op is also running on device.

LPanosTT commented:
> Please add tests in `test/ttmlir/Silicon/StableHLO/` to ensure that `dot_general` op is also running on device.

I've made a folder for the `dot_general` stablehlo silicon tests and moved the existing test in there. I've also added a new test that represents a batched matmul

uazizTT commented on lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:
```diff
@@ -279,30 +279,80 @@ class StableHLOToTTIRDotGeneralOpConversionPattern
+      return rewriter.notifyMatchFailure(
+          srcOp,
+          "LHS and RHS must have exactly 1 contracting dimension each. "
+          "Recieved LHS contracting dims: " +
```
Nit: `Received` spelling

LPanosTT replied:
Fixed

uazizTT commented on lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:
```diff
@@ -279,30 +279,80 @@ class StableHLOToTTIRDotGeneralOpConversionPattern
+
+    bool allOnes = true;
+    bool mismatchedDims = false;
+    for (int32_t i = rhsShape.size() - 3; i >= 0; i--) {
```
What if rhsShape() doesn't have 3 dimensions?

LPanosTT replied:
Then the loop won’t execute any iterations, it won’t need to because we only need to do this check if the the has 3 or more dims

uazizTT reviewed (approved):
Can you check if this passes all tt-torch tests for dot_general (and doesn't introduce new ones)?
